### PR TITLE
kcqrs: add rpc support

### LINKED ISSUE #4: kcqrs/rpc: rpc layer for basic micro-service support
Currently KCQRS is designed as library which you can plug in your project but it is capable to work only in a single app. If you are having multiple micro-services and you would like to share the same mechanism for the whole app you have to add an RPC layer over the top. 

The RPC layer need to provide client and backend implementation and this is the overall architecture which need to be supported: 

screen shot 2018-03-06 at 8 48 03

cc @mlesikov @VasilMitovClouway

### PR DESCRIPTION
**Pull request by mgenov**
Added RPC support to the library by introducing an HTTP layer over the existing EventStore implementation.

The library provides a client which is using Google HTTP Client to communicate with remote endpoint. The backend is not going to be added to the library, so you are able to make it for your-self using the
library modules such as EventStore.

Fixes #4

### CODE REVIEW

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                     private val requestFactory: HttpRequestFactory) : EventStore {
+
+    override fun saveEvents(aggregateType: String, events: List<EventPayload>, saveOptions: SaveOptions): SaveEventsResponse {
+        val aggregateId = saveOptions.aggregateId
+
+        val requestEvents = events.map { EventPayloadDto(it.kind, it.timestamp, it.identityId, it.data.payload.toString(Charsets.UTF_8)) }
+
+        try {
+            val request = requestFactory.buildPostRequest(
+                    GenericUrl(endpoint.toString() + "/v1/aggregates"),
```
Comment (VasilMitovClouway) [inline kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt:29]: Is there a reason this is hardcoded and not injected alongside the endpoint?

File: kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt [hunk 2]
```diff
@@ -3,157 +3,189 @@ package com.clouway.kcqrs.adapter.appengine
...
+
+            /*
+             * Keys of aggregates are dispatched on different levels by using an index for scalling purposes.
+             * Each aggregate in the datastore has single or multiple levels.
+             */
+            val aggregateKey = aggregateKey(aggregateId, aggregateIndex)
+            val aggregateEntity = try {
+                dataStore.get(transaction, aggregateKey)
+            } catch (ex: EntityNotFoundException) {
+                val entity = Entity(kind, "${aggregateId}_$aggregateIndex")
```
Comment (qnislavnachev) [inline kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt]: Here you can construct the entity like ``` Entity(aggregateKey) ``` instead of building the key again

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            }
+
+            return GetEventsResponse.Error("got unknown error")
+
+        } catch (ex: IOException) {
+            return GetEventsResponse.ErrorInCommunication
+        }
+    }
+
+    override fun revertLastEvents(aggregateId: String, count: Int): RevertEventsResponse {
```
Comment (VasilMitovClouway) [inline kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt:79]: Suggestion - usually it's a good practice to return the changed fields when patching.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        } catch (ex: IOException) {
+            return RevertEventsResponse.Error("Communication Error")
+        }
+    }
+
+}
+
+internal data class GetEventsResponseDto(@Key @JvmField var aggregateId: String, @Key @JvmField var aggregateType: String, @Key @JvmField var version: Long, @Key @JvmField var topic: String, @Key @JvmField var events: List<EventPayloadDto>) : GenericJson() {
+    constructor(aggregateId: String, aggregateType: String, version: Long, events: List<EventPayloadDto>) : this(aggregateId, aggregateType, version, "", events)
+    @Suppress("UNUSED")
```
Comment (VasilMitovClouway) [inline kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt]: Move this over the upper line to suppress it as well if you are going to use this.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            return RevertEventsResponse.Error("Generic Error")
+
+        } catch (ex: IOException) {
+            return RevertEventsResponse.Error("Communication Error")
+        }
+    }
+
+}
+
+internal data class GetEventsResponseDto(@Key @JvmField var aggregateId: String, @Key @JvmField var aggregateType: String, @Key @JvmField var version: Long, @Key @JvmField var topic: String, @Key @JvmField var events: List<EventPayloadDto>) : GenericJson() {
```
Comment (VasilMitovClouway) [inline kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt:106]: I'm guessing you are using the ```@Key``` and ```@JvmField``` so that when the GenericJson object is parsed we would be able to do something like ```json.get("aggregateId")``` ? Is it a requirement the fields to be var as well and not val ?

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt (hunk 4 above)
Comment (VasilMitovClouway) [reply kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt]: Or you could add default field values to escape these default constructors.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/SyncEventPublisher.kt [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * @author Miroslav Genov (miroslav.genov@clouway.com)
+ */
+class SyncEventPublisher(private val messageBus: MessageBus) : EventPublisher {
+    override fun publish(events: Iterable<EventWithPayload>) {
+        events.forEach {
+            try {
+                messageBus.handle(it)
+            } catch (ex: Exception) {
```
Comment (VasilMitovClouway) [inline kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/SyncEventPublisher.kt:13]: I think it would be better the MessageBus to throw Exceptions that extend MessageBusException for instance and to catch them here. In my opinion it's bad practice to catch all exceptions.

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/EventStore.kt [hunk 7]
```diff
@@ -3,38 +3,126 @@ package com.clouway.kcqrs.core
...
+     * @param saveOptions saving options
+     * @throws EventCollisionException is thrown in case
      */
-    @Throws(EventCollisionException::class)
-    fun saveEvents(aggregateId: UUID, expectedVersion: Int, events: Iterable<Event>)
+    fun saveEvents(aggregateType: String, events: List<EventPayload>, saveOptions: SaveOptions = SaveOptions(version = 0L)): SaveEventsResponse
 
     /**
-     * Retrieves the events
+     * Retrieves the events that are saved for the stored aggregsate.
```
Comment (VasilMitovClouway) [inline kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/EventStore.kt]: "aggregate state"

**Review by VasilMitovClouway — CHANGES_REQUESTED** [review]

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Identity.kt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package com.clouway.kcqrs.core
+
+import java.time.Instant
+
+/**
+ * Identity is representing a single identity which represents the author of the request.
+ *
+ * @author Miroslav Genov (miroslav.genov@clouway.com)
+ */
+data class Identity(val id: String, val time: Instant)
```
Comment (qnislavnachev) [inline kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Identity.kt]: In the description of the class is described that Object represents the author. I think a better name will be ``` AuthorIdentity  ``` if it`s not used for another purpose.

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/messages/MessageFormat.kt [hunk 9]
```diff
@@ -1,6 +1,7 @@
 package com.clouway.kcqrs.core.messages
 
 import java.io.InputStream
+import java.io.OutputStream
```
Comment (qnislavnachev) [inline kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/messages/MessageFormat.kt]: unused import

**Review by qnislavnachev — CHANGES_REQUESTED** [review]

**Comment by mlesikov** [comment]
>PR is not ready as there is one pending things:
>
>  provide name of the topic to which events to be published

I  think that developers can add pubsubs event publicher as an interceptor of the message bus. Like that they will have control over the topics

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt (hunk 3 above)
Comment (mgenov) [reply kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt:79]: This means that the whole aggregate root need to be returned, which doesn't sound like a good idea. I used PATCH instead of DELETE as it touches the content of the aggregate.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt (hunk 4 above)
Comment (mgenov) [reply kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt]: Actually the first constructor was used as step point which is no longer required and is now removed.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt (hunk 5 above)
Comment (mgenov) [reply kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt:106]: Yes, it's required by the parser of the google-http-client library.

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/EventStore.kt (hunk 7 above)
Comment (mgenov) [reply kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/EventStore.kt]: Changed.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/SyncEventPublisher.kt (hunk 6 above)
Comment (mgenov) [reply kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/SyncEventPublisher.kt:13]: This means that you have to force client code to throw specific exception which doesn't sound good. Keep in mind that EventHandler is implemented by the client code.

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Identity.kt (hunk 8 above)
Comment (mgenov) [reply kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Identity.kt]: Replaced author with initiator.

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/messages/MessageFormat.kt (hunk 9 above)
Comment (mgenov) [reply kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/messages/MessageFormat.kt]: Fixed.

File: kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt (hunk 2 above)
Comment (mgenov) [reply kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt]: Yes, changed to use aggregateKey instead.

File: kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt (hunk 1 above)
Comment (mgenov) [reply kcqrs-client/src/main/kotlin/com/clouway/kcqrs/client/HttpEventStore.kt:29]: Not particular reason to make it as the backend service is not scope of this project and it's strange e2e to be performed on the HTTP client.

**Comment by mgenov** [comment]
> I think that developers can add pubsubs event publicher as an interceptor of the message bus. Like that they will have control over the topics

Not really. The server backend is not able to use MessageBus and interceptors as it doesn't have java/kotlin types in it's classpath and deals only with binary messages.

**Comment by mgenov** [comment]
I though about using of the interceptor on the client, but this means that the aggregate repository should persist events in the store after the dispatching but it doesn't sound reasonable. Here is a snippet that shows what I mean:

```kotlin
val eventsWithPayload = uncommittedEvents.map { EventWithPayload(it, messageFormat.format(it)) }
        val identity = identityProvider.get()

        val events = eventsWithPayload.map {
            EventPayload(it.event::class.java.simpleName, identity.time.toEpochMilli(), identity.id, Binary(it.payload))
        }

        val response = eventStore.saveEvents(
                aggregate::class.java.simpleName,
                events,
                SaveOptions(aggregate.getId()!!, aggregate.getExpectedVersion())
        )
        when (response) {
            is SaveEventsResponse.Success -> {
                    eventPublisher.publish(eventsWithPayload)
            }
        }
    }

```

**Comment by mgenov** [comment]
The topic name is now determined depending on the aggregate. A configuration object is added which is passed as parameter of the KCQRS instance and is used to provide name of the topic associated with the provided aggregate.

**Review by mlesikov — APPROVED** [review]

File: kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineKcqrs.kt [hunk 10]
```diff
@@ -3,26 +3,51 @@ package com.clouway.kcqrs.adapter.appengine
...
+    }
+
+    class Builder(private val configuration: Configuration, private val messageFormatFactory: MessageFormatFactory) {
+        private val messageBus = SimpleMessageBus()
+        var kind: String = "Event"
+        var kcqrsHandlerEndpoint = "/worker/kcqrs"
+        var queueName: String? = null
+        var identityProvider: IdentityProvider = IdentityProvider.Default()
+        var eventStore = AppEngineEventStore(kind, messageFormatFactory.createMessageFormat())
+        var eventPublisher: EventPublisher = TaskQueueEventPublisher(kcqrsHandlerEndpoint, queueName)
```
Comment (mlesikov) [inline kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineKcqrs.kt:39]: I can't understad why we use TaskQueueEventPublisher by default

Comment (mgenov) [reply kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineKcqrs.kt:39]: Do you have something about it ? I hear from you several times that TaskQueue shouldn't be used but I didn't understood why.

Comment (mlesikov) [reply kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineKcqrs.kt:39]: I'm not sayning that we should NOT use it, In the example the events are transported throught this emssage bus (taskqueueu) - async. 
Nevermind, we will see when we start implementing the flow...

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/AuditAwareAggregateRepository.kt [hunk 11]
```diff
@@ -0,0 +1,30 @@
+package com.clouway.kcqrs.core
+
+/**
+ * Repository is representing an Repository which operates with the AggregateRoot objects.
+ *
+ * @author Miroslav Genov (miroslav.genov@clouway.com)
+ */
+interface AuditAwareAggregateRepository {
```
Comment (VasilMitovClouway) [inline kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/AuditAwareAggregateRepository.kt:8]: Name looks strange ? What is the reason for the change form AggregateRepository?

Comment (mgenov) [reply kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/AuditAwareAggregateRepository.kt:8]: The Identity information which is required for audit purposes is now moved one level up to the callstack as Configuration object was added as dependency to the AggregateRepository. 

I'm open for suggestions for better name.

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Configuration.kt [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+/**
+ * Configuration is representing the configuration of the KCQRS.
+ *
+ * @author Miroslav Genov (miroslav.genov@clouway.com)
+ */
+interface Configuration {
+    /**
+     * Determines the topic name of the AggregateRoot.
+     */
+    fun topicName(aggregate: AggregateRoot): String
```
Comment (qnislavnachev) [inline kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Configuration.kt:12]: I think we don't need to pass the whole aggregate to fetch the topic. @mgenov Do you think if we pass just ``` class<T> ``` will be enough to fetch the topic ???

**Review by qnislavnachev — APPROVED** [review]

File: kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Configuration.kt (hunk 12 above)
Comment (mgenov) [reply kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/Configuration.kt:12]: In most cases this method will return a hardcoded value of the topic name and it will be in the entry point of the micro-service. If micro-services are granular then each one will use a single topic. 

The instance may provide and additional information then the class as it contains the aggregateId version and etc. Also the check for a certain type in kotlin would be something like:

```kotlin
if (aggregate is Invoice) {
  return "invoicing"
}
```
which looks simple enough.

**Comment by mgenov** [comment]
@VasilMitovClouway I'm still waiting your approval on this PR.

**Review by VasilMitovClouway — APPROVED** [review]
Couldn't think of a better name for the AggregateRepository, I guess we can always change it in the future.

**Comment by mgenov** [comment]
Yes, name could be improved over time.
